Usernames in Gitlab are case insensative

Hi,

I'm unsure if this should be fixed or not, but it is a bit annoying. Apparently, user names in Gitlab are case insensitive, see [here](https://gitlab.com/gitlab-org/gitlab-foss/-/issues/38304):

> But usernames are unique in a case-insensitive manner, see [User model](https://gitlab.com/gitlab-org/gitlab-ce/blob/313fbd5f74c202c3a6f003a39f5ffcf583045d07/app/models/user.rb#L151). So I guess we could modify UsersFinder to make it search case-insensitive.

 Thus, for instance, the real username as is stored in GitLab is "TestUser". But you configured this user as "testuser" in your students file.

Now if you run
```repobee repos setup --tb```
with this student file, I'll get an error like:
<details>
<summary> click for details </summary>

```
[ERROR] 409: Member already exists
[ERROR] Critical exception
Traceback (most recent call last):
  File "/home/lars/.repobee/env/lib/python3.8/site-packages/gitlab/exceptions.py", line 279, in wrapped_f
    return f(*args, **kwargs)
  File "/home/lars/.repobee/env/lib/python3.8/site-packages/gitlab/mixins.py", line 214, in create
    server_data = self.gitlab.http_post(path, post_data=data, files=files, **kwargs)
  File "/home/lars/.repobee/env/lib/python3.8/site-packages/gitlab/__init__.py", line 676, in http_post
    result = self.http_request(
  File "/home/lars/.repobee/env/lib/python3.8/site-packages/gitlab/__init__.py", line 567, in http_request
    raise GitlabHttpError(
gitlab.exceptions.GitlabHttpError: 409: Member already exists

 

The above exception was the direct cause of the following exception:

 

Traceback (most recent call last):
  File "/home/lars/.repobee/env/lib/python3.8/site-packages/_repobee/ext/gitlab.py", line 80, in _try_api_request
    yield
  File "/home/lars/.repobee/env/lib/python3.8/site-packages/_repobee/ext/gitlab.py", line 178, in assign_members
    group.members.create(
  File "/home/lars/.repobee/env/lib/python3.8/site-packages/gitlab/exceptions.py", line 281, in wrapped_f
    raise error(e.error_message, e.response_code, e.response_body) from e
gitlab.exceptions.GitlabCreateError: 409: Member already exists

 

The above exception was the direct cause of the following exception:

 

Traceback (most recent call last):
  File "/home/lars/.repobee/env/lib/python3.8/site-packages/_repobee/main.py", line 174, in _main
    _repobee.cli.dispatch.dispatch_command(
  File "/home/lars/.repobee/env/lib/python3.8/site-packages/_repobee/cli/dispatch.py", line 56, in dispatch_command
    dispatch_table[category](args, config_file, api) or hook_results
  File "/home/lars/.repobee/env/lib/python3.8/site-packages/_repobee/cli/dispatch.py", line 80, in _dispatch_repos_command
    return command.setup_student_repos(
  File "/home/lars/.repobee/env/lib/python3.8/site-packages/_repobee/command/repos.py", line 83, in setup_student_repos
    platform_teams = _create_platform_teams(teams, api)
  File "/home/lars/.repobee/env/lib/python3.8/site-packages/_repobee/command/repos.py", line 115, in _create_platform_teams
    return list(
  File "/home/lars/.repobee/env/lib/python3.8/site-packages/tqdm/std.py", line 1178, in __iter__
    for obj in iterable:
  File "/home/lars/.repobee/env/lib/python3.8/site-packages/_repobee/command/teams.py", line 33, in create_teams
    api.assign_members(team, new_members, permission)
  File "/home/lars/.repobee/env/lib/python3.8/site-packages/_repobee/ext/gitlab.py", line 178, in assign_members
    group.members.create(
  File "/usr/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/lars/.repobee/env/lib/python3.8/site-packages/_repobee/ext/gitlab.py", line 93, in _try_api_request
    raise plug.PlatformError(str(e), status=e.response_code) from e
repobee_plug.exceptions.PlatformError: 409: Member already exists
[ERROR] A plugin exited with an error
```
</details>

I can ultimately trace this back to the "create_teams" function in the `command/teams.py` file.

In line 32, it says something like:
```python
existing_members = set(team.members)
new_members = set(required_team.members) - existing_members
```
It will go to this team, try to add a new member, although there is already an existing member. (but will not filter it out with the set, since that is case sensitive).



We could fix this by doing this:
```python
existing_members = {t.lower() for t in team.members}
required_members = {t.lower() for t in required_team.members}
new_members = required_members - existing_members
```

But I am not 100% sure if github ([here](https://happygitwithr.com/github-acct.html) it says it is) or gittea or whatever is also username case insensitive. Maybe you can give your thoughts.

A similar issue happens when you have a students team already setup, but it has a different case compared to what the student is called.

So suppose again you have a team named "TestStudent", but the students is configured as "teststudent"
We will get:
```
[ERROR] PlatformError: 400: Failed to save group {:path=>["has already been taken"]}
[ERROR] A plugin exited with an error
```

We can then again fix this in changing in https://github.com/repobee/repobee/blob/master/src/_repobee/ext/gitlab.py#L156 
```python
if not team_names or group.path.lower() in unique_team_names
```
And in https://github.com/repobee/repobee/blob/master/src/_repobee/command/teams.py#L21-L30
```python
    existing_teams_dict = {
        existing.name.lower(): existing
        for existing in api.get_teams({t.name.lower() for t in teams})
    }
    for required_team in teams:
        team = existing_teams_dict.get(required_team.name.lower()) or api.create_team(
            required_team.name,
            members=required_team.members,
            permission=permission,
        )
```

probably this needs to be applied for github and gittea as well.

--- a/src/repobee_plug/localreps.py
+++ b/src/repobee_plug/localreps.py
@@ -38,7 +38,7 @@ class StudentTeam:
         return self.name
 
     def __post_init__(self):
-        object.__setattr__(self, "memebers", list(self.members))
+        object.__setattr__(self, "members", [m.lower() for m in self.members])
         object.__setattr__(
             self, "name", self.name or "-".join(sorted(self.members))
         )

--- a/src/repobee_plug/platform.py
+++ b/src/repobee_plug/platform.py
@@ -62,6 +62,9 @@ class Team(APIObject):
     id: Any = dataclasses.field(compare=False)
     implementation: Any = dataclasses.field(compare=False, repr=False)
 
+    def __post_init__(self):
+        object.__setattr__(self, "members", [m.lower() for m in self.members])
+
     def __str__(self):
         return self.name
 

--- a/system_tests/gitlab/test_gitlab_system.py
+++ b/system_tests/gitlab/test_gitlab_system.py
@@ -287,6 +287,38 @@ class TestSetup:
             [plug.StudentTeam(members=[student])], assignment_names
         )
 
+    def test_setup_with_wrong_case_on_student(self, extra_args):
+        """User names are case insensitive on GitLab, and so setup should work
+        fine even if the case of some character in a student's username is
+        "incorrect".
+
+        See https://github.com/repobee/repobee/issues/900
+        """
+        student = STUDENT_TEAMS[0].members[0]
+        student_wrong_case = student.upper()
+        assert (
+            student != student_wrong_case
+        ), "cases match, test is pointless :("
+
+        command = " ".join(
+            [
+                REPOBEE_GITLAB,
+                *repobee_plug.cli.CoreCommand.repos.setup.as_name_tuple(),
+                *BASE_ARGS,
+                *TEMPLATE_ORG_ARG,
+                *MASTER_REPOS_ARG,
+                "--students",
+                student_wrong_case,
+            ]
+        )
+
+        result = run_in_docker_with_coverage(command, extra_args=extra_args)
+
+        assert result.returncode == 0
+        assert_repos_exist(
+            [plug.StudentTeam(members=student)], assignment_names
+        )
+
 
 @pytest.mark.filterwarnings("ignore:.*Unverified HTTPS request.*")
 class TestUpdate:

--- a/system_tests/gitlab/test_gitlab_system.py
+++ b/system_tests/gitlab/test_gitlab_system.py
@@ -242,6 +242,51 @@ class TestSetup:
             [plug.StudentTeam(members=[TEACHER])], assignment_names
         )
 
+    def test_setup_with_twice_with_wrong_case_on_second_setup(
+        self, extra_args
+    ):
+        """User names on GitLab are case insensitive, and so setting up repos
+        for the same student with two different cases of characters should work
+        the same as setting up just once
+
+        See
+        https://github.com/repobee/repobee/issues/900#issuecomment-830075510
+        for a bug where this was not the case.
+        """
+        student = STUDENT_TEAMS[0].members[0]
+        student_lowercase = STUDENT_TEAMS[0].members[0].lower()
+        # the original should be lowercase
+        assert (
+            student == student_lowercase
+        ), "expected real student username to be lowercase"
+
+        student_uppercase = student_lowercase.upper()
+        base_command = [
+            REPOBEE_GITLAB,
+            *repobee_plug.cli.CoreCommand.repos.setup.as_name_tuple(),
+            *BASE_ARGS,
+            *TEMPLATE_ORG_ARG,
+            *MASTER_REPOS_ARG,
+            "--students",
+        ]
+
+        def _cmd(student):
+            return " ".join(base_command + [student])
+
+        result_lowercase = run_in_docker_with_coverage(
+            _cmd(student_lowercase), extra_args=extra_args
+        )
+        result_uppercase = run_in_docker_with_coverage(
+            _cmd(student_uppercase), extra_args=extra_args
+        )
+
+        assert result_lowercase.returncode == 0
+        assert result_uppercase.returncode == 0
+
+        assert_repos_exist(
+            [plug.StudentTeam(members=[student])], assignment_names
+        )
+
 
 @pytest.mark.filterwarnings("ignore:.*Unverified HTTPS request.*")
 class TestUpdate:

--- a/system_tests/gitlab/test_gitlab_system.py
+++ b/system_tests/gitlab/test_gitlab_system.py
@@ -316,7 +316,7 @@ class TestSetup:
 
         assert result.returncode == 0
         assert_repos_exist(
-            [plug.StudentTeam(members=student)], assignment_names
+            [plug.StudentTeam(members=[student])], assignment_names
         )
 
 

--- a/tests/unit_tests/repobee_plug/test_localreps.py
+++ b/tests/unit_tests/repobee_plug/test_localreps.py
@@ -1,0 +1,13 @@
+"""Tests of the localreps classes and functions."""
+
+import repobee_plug as plug
+
+
+class TestStudentTeam:
+    """Tests for the StudentTeam class."""
+
+    def test_constructor_lowercases_member_names(self):
+        members = ["siMON", "alIce", "EVE"]
+        members_lowercase = ["simon", "alice", "eve"]
+        team = plug.StudentTeam(members=members)
+        assert team.members == members_lowercase

--- a/tests/unit_tests/repobee_plug/test_platform.py
+++ b/tests/unit_tests/repobee_plug/test_platform.py
@@ -156,3 +156,22 @@ class TestIssue:
         reconstructed = platform.Issue.from_dict(asdict)
 
         assert reconstructed == issue
+
+
+class TestTeam:
+    """Tests for the Team class."""
+
+    def test_lowercases_usernames(self):
+        """While all of the platforms currently supported are case insensitive,
+        some still allow usernames to contain e.g. capital letters. We don't
+        want this to appear in RepoBee, as it can case problems such as those
+        reported in https://github.com/repobee/repobee/issues/900.
+        """
+        usernames = ["siMon", "ALICE", "eVe"]
+        lowercase_usernames = ["simon", "alice", "eve"]
+
+        team = platform.Team(
+            members=usernames, name="yep", id=1, implementation="fake"
+        )
+
+        assert team.members == lowercase_usernames

